components/serializers.py

```diff
@@ -159,3 +160,61 @@ class Meta:
             "catalog",
             "controls_count",
         )
+
+
+class ComponentControlSerializer(serializers.ModelSerializer):
+    class Meta:
+        model = Component
+        fields = (
+            "pk",
+            "controls",
+            "description",
+            # "catalog_name",
+            "component_json",
+        )
+
+    def update(self, instance, validated_data):
+        print("inside patch function")
+        if controls := validated_data.get("controls"):
+            print(controls)
+            if controls[0] in instance.controls:
+                remove_implemented_requirement(self, instance, controls)
+            if description := validated_data.get("description"):
+                add_implemented_requirement(self, instance, controls, description)
+            instance.save()
+            return instance
+        else:
+            raise serializers.ValidationError("Controls must be provided")
+
+
+def add_implemented_requirement(self, instance, controls, description):
+    print("inside add_implemented_requirement function")
+    props = [
+        Property(name="security_control_type", value="Allocated"),
+        Property(name="provider", value="no"),
+    ]
+    ir = ImplementedRequirement(
+        props=props, control_id=controls[0], description=description
+    )
+    print("ir", ir)
+    instance.component_json.get("component-definition").get("components")[0].get(
+        "control-implementations"
+    )[0].get("implemented-requirements").append(ir)
+
```

Comment:
I think that when you append the ImplementedRequirements you need to convert it JSON since you are trying to append a Pydantic object to a JSON string. Something like `.append(json.loads(IR.json()))`